pkg/crc/machine/bundle/copier.go

```diff
@@ -0,0 +1,166 @@
+package bundle
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/code-ready/crc/pkg/compress"
+	"github.com/code-ready/crc/pkg/crc/constants"
+	"github.com/code-ready/crc/pkg/crc/logging"
+	crcos "github.com/code-ready/crc/pkg/os"
+	godeepcopy "github.com/jinzhu/copier"
+)
+
+type Copier struct {
+	srcBundle    *CrcBundleInfo
+	copiedBundle CrcBundleInfo
+}
+
+func NewCopier(srcBundle *CrcBundleInfo, basePath string, customBundleName string) (*Copier, error) {
+	var copier Copier
+
+	bundlePath := filepath.Join(basePath, customBundleName)
+	if err := os.Mkdir(bundlePath, 0775); err != nil {
+		return nil, err
+	}
+	if err := godeepcopy.Copy(&copier.copiedBundle, srcBundle); err != nil {
+		return nil, err
+	}
+
+	copier.copiedBundle.bundleName = customBundleName
+	copier.copiedBundle.Type = "custom"
+	copier.srcBundle = srcBundle
+	copier.copiedBundle.cachedPath = bundlePath
+
+	return &copier, nil
+}
+
+func (copier *Copier) Cleanup() error {
+	return os.RemoveAll(copier.copiedBundle.cachedPath)
+}
+
+func (copier *Copier) resolvePath(filename string) string {
+	return copier.copiedBundle.resolvePath(filename)
+}
+
+func (copier *Copier) CopyPrivateSSHKey(srcPath string) error {
+	sshKeyFileName := filepath.Base(copier.srcBundle.GetSSHKeyPath())
+	destPath := copier.resolvePath(sshKeyFileName)
+	if err := crcos.CopyFileContents(srcPath, destPath, 0400); err != nil {
+		return err
+	}
+	// copier.srcBundle.ClusterInfo.SSHPrivateKeyFile = sshKeyFileName
+	return nil
+}
+
+func (copier *Copier) SetKubeAdminPassword(kubePassword string) error {
+	kubeAdminPasswordFileName := copier.srcBundle.ClusterInfo.KubeadminPasswordFile
+	path := copier.resolvePath(kubeAdminPasswordFileName)
+	if err := ioutil.WriteFile(path, []byte(kubePassword), 0600); err != nil {
+		return err
+	}
+	// copier.copiedBundle.ClusterInfo.KubeadminPasswordFile = kubeAdminPasswordFileName
+	return nil
+}
+
+func (copier *Copier) CopyKubeConfig() error {
+	kubeConfigFileName := filepath.Base(copier.srcBundle.GetKubeConfigPath())
+	srcPath := copier.srcBundle.GetKubeConfigPath()
+	destPath := copier.resolvePath(kubeConfigFileName)
+	if err := crcos.CopyFileContents(srcPath, destPath, 0640); err != nil {
+		return err
+	}
+	// copier.copiedBundle.ClusterInfo.KubeConfig = kubeConfigFileName
+	return nil
+}
+
+func (copier *Copier) CopyOcExecutable() error {
+	srcPath := copier.srcBundle.GetOcPath()
+	destPath := copier.resolvePath(constants.OcExecutableName)
+	if err := crcos.CopyFileContents(srcPath, destPath, 0640); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (copier *Copier) SetDiskImage(path string) error {
+	fileInfo, err := getFileInfo(path)
+	if err != nil {
+		return err
+	}
+	diskImage := DiskImage{
+		File:   *fileInfo,
+		Format: copier.srcBundle.GetDiskImageFormat(),
+	}
+
+	copier.copiedBundle.Storage.DiskImages = []DiskImage{diskImage}
```

Comment:
You also need to update ` copier.copiedBundle.Nodes[0].DiskImage`

Reply:
https://github.com/code-ready/crc/pull/2147#discussion_r612924000